### DISCUSSION TITLE
Fantasy about async/await

### DISCUSSION THREAD

**Discussion by Aleksey-Danchin** [motion-canvas/motion-canvas#941] [General]
Guys, I played around for a couple of days. Amazing work done, and there were probably reasons for the decisions made that ultimately led to where we are. But it’s strange for me to work with generators, since I’ve long been accustomed to asynchronous functions.

I've jotted down code for an ideal approach that I'd be happy with. This of course does not mean that you need to rebuild the entire project from scratch. And it doesn’t mean that I will adapt or develop my own engine. But it would be cool if something like that could be done.

I understand how to implement everything that I wrote in async/await on generators, the same ones that are used now. Except for the case that I `marked' in the code. With dynamic addition of view points and removal of them after a short animation.

If suddenly someone shows how to implement this piece of code using generators, that would be great!

But in general, just what do you think about if motion canvas allowed you to write such code?

```ts
export default makeScene2D(async (view) => {
	console.log("start");

	const dot = (
		<Circle width={10} height={10} fill="white" position={[0, 0]} />
	);

	const rect = <Rect size={10} fill="green" position={[-300, 0]} />;

	view.add(dot);
	view.add(rect);

	// wait the end of filling
	await dot.fill("green", 10);

	// changes begin simultaneously, but both mutations are expected to end
	await dot.fill("pink", 10).x(500, 5);

	// changes begin simultaneously, but all mutations are expected to end
	await Promise.all(dot.fill("orange", 10).x(100, 5), rect.fill("black", 10));

	// chain of then-promises
	await dot
		.fill("black", 3)
		.then((dot) => {
			// dot as argument, but not used
			return rect.fill("white", 2);
		})
		.then((rect) => {
			// rect as argument, but used
			rect.x(0, 1);
		});

	// Here you will need to strain your brains.
	// <marked>
	const promises: Promise<any>[] = [];
	// I understand that there is a dependence on fps
	await tween(5, async (value) => {
		dot.x(map(-800, 800, value));

		// 60fps * 5sec = 300 particles (It's ok)
		const point = <Circle position={dot.position()} width={5} height={5} />;

		view.add(point);

		// wait for the start of hiding 5 seconds for each particle
		const promise = waitFor(5).then(async () => {
			await point.opacity(0, 3);
			point.remove();
		});

		promises.push(promise);
	});

	// waiting for all the particles to disappear
	await Promise.all(promises);
	// </marked>

	console.log("done");
});
```

**Comment by Aleksey-Danchin** [comment]
In general, the idea with async/await is not so bad. Without animation, but with the general concept of working with frames.

https://github.com/Aleksey-Danchin/AnotherOneMotion

**Comment by aarthificial** [comment]
Generators work better with animation because you can explicitly progress them each frame by calling `next` and cancel them by calling `return`. No additional work needed, that's just how they work.

With `async/await` you'd need to extract the `resolve` callback from every possible promise so that you can manually call it to progress the animation, and that's assuming that people wouldn't write their own promises. Your implementation doesn't seem to do that because it only logs things to the console. Normally you'd need to synchronize the frames with real-time and/or audio clock and account for the user pausing and restarting the playback.

This is how you can implement your example in Motion Canvas right now:
```ts
const tasks = [];
yield dot().x(-800).x(800, 5);
while (dot().x.context.isTweening()) {
  const point = (
    <Circle position={dot().position()} size={5} fill={'white'} />
  );
  view.add(point);
  tasks.push(yield delay(5, point.opacity(0, 3)));
  yield;
}

yield* join(...tasks);
```
But it highlights a missing piece in our implementation, namely the `spawn` function.
Right now it's only possible to spawn a new thread by `yield`ing it, with a spawn function you can closely replicate what you are doing:
```ts
const tasks = [];
yield* tween(5, value => {
  dot().x(map(-800, 800, value));
  const point = (
    <Circle position={dot().position()} size={5} fill={'white'} />
  );
  view.add(point);
  tasks.push(spawn(delay(5, point.opacity(0, 3))));
});

yield* join(...tasks);
```

> But in general, just what do you think about if motion canvas allowed you to write such code?

The core of Motion Canvas is generator-independent. If you can implement the `Scene` interface, Motion Canvas can run it.
`GeneratorScene` used by our 2D package is just one of the concrete implementations. 
I'm personally not interested in implementing any `async/await`-based solutions due to the reasons stated above.
